airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_incremental.py

```diff
@@ -180,24 +245,35 @@ def test_two_sequential_reads(
             latest_state = states_1[-1].state.data
             state_input = states_1[-1].state.data
 
-        parsed_records_1 = list(records_with_state(records_1, latest_state, stream_mapping, cursor_paths))
-
-        # This catches the case of a connector that emits an invalid state that is not compatible with the schema
-        # See https://github.com/airbytehq/airbyte/issues/21863 to understand more
-        assert parsed_records_1, "At least one valid state should be produced, given a cursor path"
-
-        for record_value, state_value, stream_name in parsed_records_1:
-            assert (
-                record_value <= state_value
-            ), f"First incremental sync should produce records younger or equal to cursor value from the state. Stream: {stream_name}"
+        # READ #2
 
         output = docker_runner.call_read_with_state(connector_config, configured_catalog_for_incremental, state=state_input)
         records_2 = filter_output(output, type_=Type.RECORD)
+        records_by_stream_2 = group_records_by_stream(records_2)
+
+        assert records_2, "Second Read should produce at least one record"
 
-        for record_value, state_value, stream_name in records_with_state(records_2, latest_state, stream_mapping, cursor_paths):
-            assert compare_cursor_with_threshold(
```

Comment:
Copied from PR Description:

---

**1. Cursors are not always defined and/or comparable**
We have been treating ALL cursors as comparable and defined.

This is not true.

Some connectors (e.g. source-postgres) uses a source-defined cursor in many cases which is opaque to our system.

As a result in these cases we cannot test if records emitted are greater than or less than the cursor value.

Also not all cursors are comparable! A connector could know how to compare two cursor values when we dont.

For example a valid cursor field could be a string which represents an concatingated field foo which has a value unixdatetime:123453;someotherdata:bar

Reply:
I agree with your statement "Also not all cursors are comparable", heck they could be a bitmap or something like that.  That said, I do think we should be testing that if we replay a sync with the same state it just gave us, only the "latest" record should emitted, or no new records.  Is that this test, or another test?